src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -0,0 +1,310 @@
+{
+  goto_functionst &goto_functions = goto_model.goto_functions;
+  symbol_tablet &symbol_table = goto_model.symbol_table;
+
+  goto_functiont &goto_function = goto_functions.function_map[function];
+  const symbolt &function_symbol = symbol_table.lookup_ref(function);
+
+  goto_programt &goto_program = goto_function.body;
+
+  const symbolt &symbol = get_fresh_aux_symbol(
+    bool_typet(),
+    id2string(function),
+    "func_init_done",
+    source_locationt(),
+    function_symbol.mode,
+    symbol_table);
+
+  const symbol_exprt &var = symbol.symbol_expr();
+
+  // initialise var in __CPROVER_initialize if it is present
+  const auto f_it =
+    goto_functions.function_map.find(CPROVER_PREFIX "initialize");
+
+  if(f_it != goto_functions.function_map.end())
+  {
+    goto_programt &goto_program = f_it->second.body;
+
+    auto init_it =
+      goto_program.insert_before(std::prev(goto_program.instructions.end()));
+
+    init_it->make_assignment(code_assignt(var, false_exprt()));
+  }
+
+  const goto_programt::const_targett start_it =
+    goto_program.instructions.begin();
+
+  goto_programt::const_targett it;
+
+  if(location_number.has_value())
+  {
+    const auto opt_it = goto_program.get_target(*location_number);
+
+    if(!opt_it.has_value())
+    {
+      throw invalid_command_line_argument_exceptiont(
+        "no instruction with location number " +
+        std::to_string(*location_number) + " in function " + function);
+    }
+
+    it = *opt_it;
+  }
+  else
+  {
+    it = start_it;
+  }
+
+  // the location of where to start executing (needs to be retrieved before
+  // modifying the goto model)
+  goto_programt::const_targett it =
+    location_number.has_value() ? goto_program.get_target(*location_number)
+                                : start_it;
+
+  auto ins_it1 = goto_program.insert_before(start_it);
+  ins_it1->make_goto(goto_program.const_cast_target(start_it));
+  ins_it1->guard = var;
+
+  auto ins_it2 = goto_program.insert_after(ins_it1);
+  ins_it2->make_assignment(code_assignt(var, true_exprt()));
+
+  auto ins_it3 = goto_program.insert_after(ins_it2);
+  ins_it3->make_goto(goto_program.const_cast_target(it));
+}
+
+void memory_snapshot_harness_generatort::add_assignments_to_globals(
+  const symbol_tablet &snapshot,
+  code_blockt &code) const
+{
+  for(const auto &pair : snapshot)
+  {
+    const symbolt &symbol = pair.second;
+
+    if(!symbol.is_static_lifetime)
+      continue;
+
+    code_assignt code_assign(symbol.symbol_expr(), symbol.value);
+
+    code.add(code_assign);
+  }
+}
+
+void memory_snapshot_harness_generatort::add_call_with_nondet_arguments(
+  const symbolt &called_function_symbol,
+  code_blockt &code) const
+{
+  const code_typet &code_type = to_code_type(called_function_symbol.type);
+
+  code_function_callt::argumentst arguments;
+
+  for(const code_typet::parametert &parameter : code_type.parameters())
+  {
+    arguments.push_back(side_effect_expr_nondett(parameter.type()));
```

Comment:
This constructor of `side_effect_expr_nondett` is deprecated, please use the one that also take as source location (just use `called_function_symbol.location`).

Reply:
Done.